// parse-server/lib/GraphQL/loaders/configMutations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateCloudConfig = exports.load = void 0;
var _graphql = require("graphql");
var _graphqlRelay = require("graphql-relay");
var _node = _interopRequireDefault(require("parse/node"));
var _Error = require("../../Error");
var _GlobalConfigRouter = _interopRequireDefault(require("../../Routers/GlobalConfigRouter"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const globalConfigRouter = new _GlobalConfigRouter.default();
const updateCloudConfig = async (context, paramName, value, isMasterKeyOnly = false) => {
  const {
    config,
    auth
  } = context;
  if (!auth.isMaster) {
    throw (0, _Error.createSanitizedError)(_node.default.Error.OPERATION_FORBIDDEN, 'Master Key is required to update GlobalConfig.');
  }
  await globalConfigRouter.updateGlobalConfig({
    body: {
      params: {
        [paramName]: value
      },
      masterKeyOnly: {
        [paramName]: isMasterKeyOnly
      }
    },
    config,
    auth,
    context
  });
  return {
    value,
    isMasterKeyOnly
  };
};
exports.updateCloudConfig = updateCloudConfig;
const load = parseGraphQLSchema => {
  const updateCloudConfigMutation = (0, _graphqlRelay.mutationWithClientMutationId)({
    name: 'UpdateCloudConfig',
    description: 'Updates the value of a specific parameter in GlobalConfig.',
    inputFields: {
      paramName: {
        description: 'The name of the parameter to set.',
        type: new _graphql.GraphQLNonNull(_graphql.GraphQLString)
      },
      value: {
        description: 'The value to set for the parameter.',
        type: new _graphql.GraphQLNonNull(_graphql.GraphQLString)
      },
      isMasterKeyOnly: {
        description: 'Whether this parameter should only be accessible with master key.',
        type: _graphql.GraphQLBoolean,
        defaultValue: false
      }
    },
    outputFields: {
      cloudConfig: {
        description: 'The updated config value.',
        type: new _graphql.GraphQLNonNull(parseGraphQLSchema.cloudConfigType)
      }
    },
    mutateAndGetPayload: async (args, context) => {
      try {
        const {
          paramName,
          value,
          isMasterKeyOnly
        } = args;
        const result = await updateCloudConfig(context, paramName, value, isMasterKeyOnly);
        return {
          cloudConfig: result
        };
      } catch (e) {
        parseGraphQLSchema.handleError(e);
      }
    }
  });
  parseGraphQLSchema.addGraphQLType(updateCloudConfigMutation.args.input.type.ofType, true, true);
  parseGraphQLSchema.addGraphQLType(updateCloudConfigMutation.type, true, true);
  parseGraphQLSchema.addGraphQLMutation('updateCloudConfig', updateCloudConfigMutation, true, true);
};
exports.load = load;
