// parse-server/lib/GraphQL/loaders/configQueries.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.load = exports.cloudConfig = void 0;
var _graphql = require("graphql");
var _node = _interopRequireDefault(require("parse/node"));
var _Error = require("../../Error");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const cloudConfig = async (context, paramName) => {
  const {
    config,
    auth
  } = context;
  if (!auth.isMaster) {
    throw (0, _Error.createSanitizedError)(_node.default.Error.OPERATION_FORBIDDEN, 'Master Key is required to access GlobalConfig.');
  }
  const results = await config.database.find('_GlobalConfig', {
    objectId: '1'
  }, {
    limit: 1
  });
  if (results.length !== 1) {
    return {
      value: null,
      isMasterKeyOnly: null
    };
  }
  const globalConfig = results[0];
  const params = globalConfig.params || {};
  const masterKeyOnly = globalConfig.masterKeyOnly || {};
  if (params[paramName] !== undefined) {
    return {
      value: params[paramName],
      isMasterKeyOnly: masterKeyOnly[paramName] ?? null
    };
  }
  return {
    value: null,
    isMasterKeyOnly: null
  };
};
exports.cloudConfig = cloudConfig;
const load = parseGraphQLSchema => {
  if (!parseGraphQLSchema.cloudConfigType) {
    const cloudConfigType = new _graphql.GraphQLObjectType({
      name: 'ConfigValue',
      fields: {
        value: {
          type: _graphql.GraphQLString
        },
        isMasterKeyOnly: {
          type: _graphql.GraphQLBoolean
        }
      }
    });
    parseGraphQLSchema.addGraphQLType(cloudConfigType, true, true);
    parseGraphQLSchema.cloudConfigType = cloudConfigType;
  }
  parseGraphQLSchema.addGraphQLQuery('cloudConfig', {
    description: 'Returns the value of a specific parameter from GlobalConfig.',
    args: {
      paramName: {
        type: new _graphql.GraphQLNonNull(_graphql.GraphQLString)
      }
    },
    type: new _graphql.GraphQLNonNull(parseGraphQLSchema.cloudConfigType),
    async resolve(_source, args, context) {
      try {
        return await cloudConfig(context, args.paramName);
      } catch (e) {
        parseGraphQLSchema.handleError(e);
      }
    }
  }, false, true);
};
exports.load = load;
